s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/S3SourceTask.java

```diff
@@ -107,8 +110,9 @@ public void start(final Map<String, String> props) {
         initializeConverters();
```

Comment:
It's a little unusual/inflexible to instantiate the key and value converter in the task.

For a source task, the converters are for serializing the data into Kafka, and that can be independent of your input format.

The way we're instantiating the class directly also leaves them unconfigured, which may be undefined/undesirable for some converters.

For example, imagine someone configured this task with an Avro value.converter. Then their input data needs to already be in avro format, and has to be readable with an unconfigured Avro converter. That sounds like a very rigid setup.

Reply:
As far as I can tell, the data flow is (solid), and should be (dotted)
```mermaid
flowchart TB
InputStream -- Transformer.getRecords -->  Object -- Transformer.getValueBytes --> firstByte["byte[]"] -- Converter.toConnectData --> firstStruct[Struct] -- SMTs --> secondStruct[Struct] -- Converter.fromConnectData --> secondByte["byte[]"] -- Producer --> Kafka

InputStream -. "Transformer.getRecords" .-> firstStruct
```

Reply:
So the getValueBytes transforms(deserializes) an avro record and the RecordProcessor converts these bytes to AvroRecord, if avro converter is configured.
If no converter is configured, default is bytearrayconverter and just bytes are returned, in which case, Consumer has to serialize them back to GenericRecord if they want to read as Avro records.

May be I didn't understand your point totally.

We can move these converters initialization to record processor too.

Reply:
> We can move these converters initialization to record processor too.

My point was not where in the task the key/value converters are instantiated; they shouldn't be instantiated at all.

Your data format in S3 should be decoupled from your data format in Kafka.
e.g. Someone has Avro or parquet or protobuf or whatever data in S3, and wants to write Json to Kafka.

How would you configure that in the current implementation? `value.converter=JsonConverter` and `input.format=avro` would pass avro-serialized data from getValueBytes to the JsonConverter, which would throw an exception.

Reply:
Will be handled in ticket https://aiven.atlassian.net/browse/KCON-25